test/framework/toolchain.py

```diff
@@ -381,6 +382,68 @@ def test_optarch_aarch64_heuristic(self):
         self.assertEqual(tc.options.options_map['optarch'], 'mcpu=cortex-a72.cortex-a53')
         self.assertTrue('-mcpu=cortex-a72.cortex-a53' in os.environ['CFLAGS'])
         self.modtool.purge()
+  
+    def test_compiler_dependent_optarch(self):
+        """Test whether specifying optarch on a per compiler basis works."""
+        flag_vars = ['CFLAGS', 'CXXFLAGS', 'FCFLAGS', 'FFLAGS', 'F90FLAGS']
+        intel_options = [('intelflag','intelflag'),('GENERIC','xSSE2'),('','')]
+        gcc_options = [('gccflag','gccflag'),('GENERIC','march=x86-64 -mtune=generic'),('','')]
+        toolchains = [('iccifort','2011.13.367'),('GCC','4.7.2'),('PGI','16.7-GCC-5.4.0-2.26')]
+        enabled = [True, False]
+
+        i = product(intel_options, gcc_options, toolchains, enabled)
+
+        for (intelflags, intel_expanded_flags), (gccflags, gcc_expanded_flags), (toolchain, toolchain_version), enable in i:
+            optarch_var = {} 
+            optarch_var['Intel'] = '%s' % intelflags
+            optarch_var['GCC'] = '%s' % gccflags
+            build_options = {'optarch': optarch_var}
+            init_config(build_options=build_options)
+            tc = self.get_toolchain(toolchain, version=toolchain_version)
+            tc.set_options({'optarch': enable})
+            tc.prepare()
+            flag = None
+            if toolchain == 'iccifort':
+                flag = intel_expanded_flags
+            elif toolchain == 'GCC':
+                flag = gcc_expanded_flags 
+            else: # PGI as an example of compiler not set
+                # default optarch flag, should be the same as the one in
+                # tc.COMPILER_OPTIMAL_ARCHITECTURE_OPTION[(tc.arch,tc.cpu_family)]
+                flag = ''
+            
+            flags = tc.options.options_map['optarch']
+
+            self.assertTrue(flag == flags, "optarch: True means '%s' == '%s'" % (flag, flags))
+            
+            # Also check that it is correctly passed to xFLAGS, honoring 'enable'
+            if flag == '':
+                blacklist = [
+                       intel_options[0][1],
+                       intel_options[1][1],
+                       gcc_options[0][1],
+                       gcc_options[1][1],
```

Comment:
hmm, would it be clearer to just hardcode those strings here?

hardcoding is fine in tests if it makes things clearer